### ISSUE TITLE
[neptune] AssociatedRoles for CfnDBCluster

### ISSUE THREAD

**Issue by namedgraph** [aws/aws-cdk#10224]
`AssociatedRoles` is [now supported by CF templates for Amazon Neptune](https://github.com/awslabs/aws-cloudformation-templates/issues/153), but not available in the CDK (as of v1.62.0).

### Use Case

Associating an IAM role with a Neptune's `CfnDBCluster`.

### Proposed Solution

Should be possible as `AssociatedRoles` is [available for other types of `DBCluster`](https://docs.aws.amazon.com/cdk/api/latest/docs/@aws-cdk_aws-rds.CfnDBCluster.html).

### Other

* [ ] :wave: I may be able to implement this feature request
* [ ] :warning: This feature might incur a breaking change

---

This is a :rocket: Feature Request

**Comment by njlynch** [comment]
This is being done as part of #10201. Closing this out as a duplicate.

**Comment by namedgraph** [comment]
@njlynch so when is this going to be released?

I see `CfnDBCluster::setAssociatedRoles` in [the documentation](https://docs.aws.amazon.com/cdk/api/latest/java/index.html?software/amazon/awscdk/services/neptune/CfnDBCluster.html), but the method is not available in the latest CDK version 1.63.0.

**Comment by njlynch** [comment]
@namedgraph  - This was released in 1.64.0 (released 2020-09-22).

**Comment by namedgraph** [comment]
Thanks!

**Comment by namedgraph** [comment]
@njlynch I'm getting circular reference error when attempting to assign an associated role to a Neptune cluster:
```java
Caused by: software.amazon.jsii.JsiiException: Resolution error: Resolution error: Unable to resolve object tree with circular reference. Path:
/Resources/${Token[CdkTestStack.NeptuneCluster.LogicalID.157]}/Properties/associatedRoles/0/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host..
```
What could be the problem?

Role code:
```java
Role bulkLoaderRole = new Role(this, "RDFBulkLoaderRole", RoleProps.builder().
        assumedBy(new ServicePrincipal("rds.amazonaws.com")).
        build());
bulkLoaderRole.addToPolicy(PolicyStatement.Builder.create().
        effect(Effect.ALLOW).
        actions(Arrays.asList("s3:Get*", "s3:List*")).
        resources(Arrays.asList("*")).
        build());
```
Cluster code:
```
return CfnDBCluster.Builder.create(this, "NeptuneCluster").
        dbSubnetGroupName(subnetGroup.getDbSubnetGroupName()).
        vpcSecurityGroupIds(securityGroups).
        associatedRoles(Arrays.asList(associatedRole)).
        dbClusterIdentifier("OctopusTriplestoreCluster").
        iamAuthEnabled(Boolean.TRUE).
        deletionProtection(Boolean.TRUE).
        engineVersion("1.0.2.2").
        availabilityZones(getAvailabilityZones()).
        port(8182).
        deletionProtection(false).
        //enableCloudwatchLogsExports(securityGroups).
        build();
```

**Comment by namedgraph** [comment]
@njlynch I'm getting circular reference error when attempting to assign an associated role to a Neptune cluster:
```java
Caused by: software.amazon.jsii.JsiiException: Resolution error: Resolution error: Unable to resolve object tree with circular reference. Path:
/Resources/${Token[CdkTestStack.NeptuneCluster.LogicalID.157]}/Properties/associatedRoles/0/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host/node/host..
```
What could be the problem?

Role code:
```java
Role bulkLoaderRole = new Role(this, "RDFBulkLoaderRole", RoleProps.builder().
        assumedBy(new ServicePrincipal("rds.amazonaws.com")).
        build());
bulkLoaderRole.addToPolicy(PolicyStatement.Builder.create().
        effect(Effect.ALLOW).
        actions(Arrays.asList("s3:Get*", "s3:List*")).
        resources(Arrays.asList("*")).
        build());
```
Cluster code:
```
return CfnDBCluster.Builder.create(this, "NeptuneCluster").
        dbSubnetGroupName(subnetGroup.getDbSubnetGroupName()).
        vpcSecurityGroupIds(securityGroups).
        associatedRoles(Arrays.asList(associatedRole)).
        dbClusterIdentifier("OctopusTriplestoreCluster").
        iamAuthEnabled(Boolean.TRUE).
        deletionProtection(Boolean.TRUE).
        engineVersion("1.0.2.2").
        availabilityZones(getAvailabilityZones()).
        port(8182).
        deletionProtection(false).
        //enableCloudwatchLogsExports(securityGroups).
        build();
```

**Comment by njlynch** [comment]
> ```
> Caused by: software.amazon.jsii.JsiiException: Resolution error: Resolution error: Unable to resolve object tree with circular reference. Path:
> /Resources/${Token[CdkTestStack.NeptuneCluster.LogicalID.157]}/Properties/associatedRoles/0/node/host/node/host/node/host/node/host/...
> ```

That repeating '/node/host/node/host' is certainly suspect. I'm not seeing anything from your code above. I suspect the issue lies somewhere in the translation from `bulkLoaderRole` to `associatedRole` in the cluster code. Do you have any constructs/nodes named 'host' there, and/or referencing the 'host' property of another construct?

**Comment by namedgraph** [comment]
The only related code that I can see is the following -- this I commented out, but still get the same error:

```java
//        String neptuneHost = Token.asString("https://" + neptune.getAttrEndpoint() + ":" + neptune.getAttrPort());

        Function bulkLoader = bulkLoader(vpc, neptuneSg);
//        bulkLoader.addEnvironment("ENDPOINT_URI", neptuneHost).
//                addEnvironment("FORMAT", "ntriples").
//                addEnvironment("IAM_ROLE_ARN", bulkLoaderRole.getRoleArn()).
//                addEnvironment("REGION", getRegion());
```

**Comment by namedgraph** [comment]
I can try to make a minimal case if that helps. So far the only way to get rid of the error is to remove the `associatedRoles()` call from the cluster.

**Comment by njlynch** [comment]
Sure, a minimal case would be useful.

**Comment by namedgraph** [comment]
@njlynch here you go: https://github.com/namedgraph/neptune-cdk-test/blob/master/src/main/java/com/myorg/NeptuneCdkTestStack.java

I hope it's enough to demonstrate the problem. Very possibly something's wrong in my code, but I just cannot spot it.

**Comment by namedgraph** [comment]
@njlynch can you confirm?

**Comment by njlynch** [comment]
Ah, I see it now. 

tl;dr - Here's the fix to your code:

```java
CfnDBCluster.DBClusterRoleProperty clusterRole = new CfnDBCluster.DBClusterRoleProperty.Builder().
    roleArn(associatedRole.getRoleArn()).
    build();

return CfnDBCluster.Builder.create(this, "NeptuneCluster").
    // ...
    associatedRoles(Arrays.asList(clusterRole)).
    // ...
    .build();
```

The associated roles isn't just an IAM Role, it's a specific type expected by CloudFormation: https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-neptune-dbcluster.html#cfn-neptune-dbcluster-associatedroles

The Java docs are not helpful here, showing the type of associatedRoles as `java.util.List<? extends java.lang.Object>`. 
https://docs.aws.amazon.com/cdk/api/latest/java/software/amazon/awscdk/services/neptune/CfnDBCluster.Builder.html#associatedRoles-java.util.List- . What it _should_ be is `java.util.List<CfnDBCluster.DBClusterRoleProperty>`.
https://docs.aws.amazon.com/cdk/api/latest/java/software/amazon/awscdk/services/neptune/CfnDBCluster.DBClusterRoleProperty.html

If you look at the Typescript or Python docs, you'll see slightly more helpful types:
* [Typescript](https://docs.aws.amazon.com/cdk/api/latest/docs/@aws-cdk_aws-neptune.CfnDBClusterProps.html#associatedroles) - `Type: IResolvable | DBClusterRoleProperty | IResolvable[]`
* [Python](https://docs.aws.amazon.com/cdk/api/latest/python/aws_cdk.aws_neptune/CfnDBClusterProps.html#aws_cdk.aws_neptune.CfnDBClusterProps.associated_roles) - `Union[IResolvable, List[Union[DBClusterRoleProperty, IResolvable]], None]`

I am really intrigued by the bad error message you got, and by the poor Java docs in this example. I'll follow up on our jsii project -- responsible for both -- and see if we can't make some improvements.

Either way, hopefully this gets you up and running now.

**Comment by RomainMuller** [comment]
This is type-union-induced, hence this RFC is relevant: https://github.com/aws/aws-cdk-rfcs/issues/193

**Comment by namedgraph** [comment]
@njlynch thanks, the error went away, but I'm still struggling to the a lambda to bulk load data into Neptune.

Could we please get Java examples as well?
https://docs.aws.amazon.com/neptune/latest/userguide/get-started-cfn-lambda.html

**Comment by njlynch** [comment]
I'm closing this out, as it seems to be resolved; feel free to comment/re-open if you have any other follow-ups. The fixes to the messaging and docs are being tracked as part of the above RFC.
